docs/source/csdl-spec.md

```diff
@@ -0,0 +1,630 @@
+      Composer
+    end
+    subgraph csdl
+      CSDL
+    end
+    subgraph Consumer
+      Router
+    end    
+    A-->Composer
+    B-->Composer
+    C-->Composer
+    Composer-->CSDL
+    CSDL-->Router
+    Router-->Clients
+    Clients-->Router
+```
+
+- **Producer** the bit that generates CSDL. spec will place requirements on composers.
+- **Consumer** the bit that consumes CSDL. spec will place requirements on consumers.
+- **Composer** a kind of producer which composes subgraph schemas into CSDL. (no requirements on consumers, diagrammed for clarity.)
+- **Subgraphs** subgraphs which are composed by the composer (no requirements on subgraphs, diagrammed for clarity.)
+- **Router** a kind of consumer which serves CSDL as a graphql endpoint. no requirements on routers, these are described in the federation spec.
+
+## Structure
+
+A CSDL document must be a valid [GraphQL schema definition language](https://spec.graphql.org/draft/#sec-Type-System) document.
+
+CSDL removes certain SDL features, as described in the [divergence](#divergence-from-graphql-sdl) section. Nevertheless, CSDL consumers **should** be prepared to parse any valid GraphQL SDL. A CSDL document which contains disallowed SDL constructs **should** trigger validation errors after parsing.
+
+CSDL introduces a suite of types and directives. All CSDL-specific types and directives are prefixed with `csdl_` as recommended in [the GraphQL specification](https://spec.graphql.org/draft/#note-ca863).
+
+### Divergence from GraphQL SDL
+Formally, CSDL is a subset of the [GraphQL schema definition language](https://spec.graphql.org/draft/#sec-Type-System).
+
+In particular, CSDL differs from GraphQL SDL in the following ways:
+- **Extensions are forbidden.** You cannot `extend type` or `extend interface` within a CSDL document.
+- **`schema @csdl(version:)` is mandatory.** Conforming CSDL documents must have a `schema` declaration which must be annotated with a `@csdl` directive specifying, at a minimum, the semver for the CSDL spec in use.
+- the CSDL must define certain types as described in the [mandatory types](#mandatory-types) section.
+
+### Mandatory Types
+
+A CSDL document **must** define a `csdl_Graph` enum. Each enum value defines a subgraph. Each value **must** be annotated with a `@csdl_endpoint` directive specifying the endpoint URL for the subgraph.
+
+```graphql
+enum csdl_Graph {
+  ROCKETS @csdl_endpoint(url: "https://rockets.api.com"),
+  ASTRONAUTS @csdl_endpoint(url: "https://astronauts.api.com"),
+}
+```
+
+The `csdl_Graph` enum is used as input to csdl directives which link fields and types to subgraphs.
+
+## Scalars
+
+### `scalar csdl_SelectionSet`
+
+A GraphQL selection set represented with the same syntax as defined in [the GraphQL SDL](https://spec.graphql.org/draft/#sec-Selection-Sets).
+
+Note: Unlike the federation `_FieldSet` scalar, a `csdl_SelectionSet` must parse as a GraphQL selection set, so the selections must be surrounded by braces.
```

Comment:
It might be worthwhile explaining the rationale behind such a discrepancy. Would we consider adding support for that format in federation as well & deprecating the current non-bracketed (or vice versa) for reasons of consistency?

Reply:
So the reason the braces were introduced in the first place is that it was annoying (and introduced a copy, so some small perf impact) to add `{` and `}` to the `_FieldSet` in order to get it to parse as a [SelectionSet](https://spec.graphql.org/draft/#sec-Selection-Sets), which is specified as including the braces.

That's obviously not enough of a reason on its own, but the thing is, `_FieldSet`s don't actually correspond to any rule within the GraphQL query grammar, so (1) it's a little harder to specify, and (2) it's likely that many other implementers will also run into exactly the same issue we did in their own language's parsers.

I'd definitely consider deprecating `_FieldSet` within federation, but gosh what an annoying change for such small gains.